model/jpa/src/main/java/org/keycloak/models/jpa/RealmAdapter.java

```diff
@@ -1777,55 +1763,34 @@ public void removeComponent(ComponentModel component) {
         if (c == null) return;
         session.users().preRemove(this, component);
         removeComponents(component.getId());
-        em.createNamedQuery("deleteComponentConfigByComponent").setParameter("component", c).executeUpdate();
-        em.remove(c);
+        getEntity().getComponents().remove(c);
     }
 
     @Override
     public void removeComponents(String parentId) {
-        TypedQuery<String> query = em.createNamedQuery("getComponentIdsByParent", String.class)
-                .setParameter("realm", realm)
-                .setParameter("parentId", parentId);
-        List<String> results = query.getResultList();
-        if (results.isEmpty()) return;
-        for (String id : results) {
-            session.users().preRemove(this, getComponent(id));
-        }
-        em.createNamedQuery("deleteComponentConfigByParent").setParameter("parentId", parentId).executeUpdate();
-        em.createNamedQuery("deleteComponentByParent").setParameter("parentId", parentId).executeUpdate();
-
+        getEntity().getComponents().stream()
+                .filter(c -> parentId.equals(c.getParentId()))
+                .forEach(c -> session.users().preRemove(this, getComponent(c.getId())));
```

Comment:
Thanks, that looks good though I'd lean towards something more fluent:

```java
@Override
public void removeComponents(String parentId) {
    getEntity().getComponents().stream()
            .filter(c -> Objects.equals(parentId, c.getParentId()))
            .forEach(c -> {
                session.users().preRemove(this, getComponent(c.getId()));
                getEntity().getComponents().remove(c);
            });
}
```

I'd need to verify it really works though as time will permit, most likely tomorrow.